topology/probes/k8s/container.go

```diff
@@ -0,0 +1,171 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"sync"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+
+	"k8s.io/api/core/v1"
+)
+
+type containerCache struct {
+	sync.RWMutex
+	defaultKubeCacheEventHandler
+	graph.DefaultGraphListener
+	*kubeCache
+	graph            *graph.Graph
+	podIndexer       *graph.MetadataIndexer
+	containerIndexer *graph.MetadataIndexer
+}
+
+// commonly accessed docker specific fields
+const (
+	DockerNameField         = "Docker.ContainerName"
+	DockerPodNamespaceField = "Docker.Labels.io.kubernetes.pod.namespace"
+	DockerPodNameField      = "Docker.Labels.io.kubernetes.pod.name"
+)
+
+func addNotNullFilter(m graph.Metadata, field string) {
+	m[field] = filters.NewNotFilter(filters.NewNullFilter(field))
+}
+
+func newContainerIndexer(g *graph.Graph) *graph.MetadataIndexer {
+	m := graph.Metadata{"Type": "container"}
+	addNotNullFilter(m, DockerPodNamespaceField)
+	addNotNullFilter(m, DockerPodNameField)
+
+	return graph.NewMetadataIndexer(g, m, DockerPodNamespaceField, DockerPodNameField)
+}
+
+func (c *containerCache) newMetadata(pod *v1.Pod, container *v1.Container) graph.Metadata {
+	m := newMetadata("container", container.Name, container)
+	common.SetField(m, DockerNameField, container.Name)
+	common.SetField(m, DockerPodNamespaceField, pod.GetNamespace())
+	common.SetField(m, DockerPodNameField, pod.GetName())
+	return m
+}
+
+func makeContainerUID(pod *v1.Pod, containerName string) string {
+	return string(pod.GetUID()) + "-" + containerName
+}
+
+func (c *containerCache) linkContainerToPod(pod *v1.Pod, container *v1.Container, containerNode *graph.Node) {
+	podNodes := c.podIndexer.Get(pod.GetName())
+
+	if len(podNodes) == 0 {
+		logging.GetLogger().Warningf("Can't find pod{%s}", container.Name, pod.GetName())
+		return
+	}
+
+	logging.GetLogger().Infof("Linking container{%s} to pod{%s}", container.Name, pod.GetName())
+	topology.AddOwnershipLink(c.graph, podNodes[0], containerNode, nil)
+}
+
+func (c *containerCache) onContainerAdd(pod *v1.Pod, container *v1.Container) {
+	uid := makeContainerUID(pod, container.Name)
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.graph.Lock()
+	defer c.graph.Unlock()
+
+	containerNode := c.graph.GetNode(graph.Identifier(uid))
+	if containerNode == nil {
+		logging.GetLogger().Infof("Creating container{%s}", container.Name)
+		containerNode = c.graph.NewNode(graph.Identifier(uid), c.newMetadata(pod, container))
+	} else {
+		logging.GetLogger().Infof("container{%s} already exists", container.Name)
+		addMetadata(c.graph, containerNode, container)
```

Comment:
I'm sure we should copy all the container metadata directly since we are going to move more or less to something schema based, but I think that is acceptable for now.